src/status.ts

```diff
@@ -0,0 +1,267 @@
+import * as vscode from "vscode";
+
+import {
+  isGemMissing,
+  addGem,
+  bundleCheck,
+  bundleInstall,
+  isGemOutdated,
+  updateGem,
+} from "./bundler";
+import { Ruby } from "./ruby";
+
+export const enum ServerCommand {
+  Start = "ruby-lsp.start",
+  Stop = "ruby-lsp.stop",
+  Error = "ruby-lsp.restart",
+}
+
+export class StatusItem {
+  private context: vscode.ExtensionContext;
+  private ruby: Ruby;
+  private selector: vscode.DocumentSelector;
+  private serverStatus: vscode.LanguageStatusItem;
+
+  constructor(context: vscode.ExtensionContext, ruby: Ruby) {
+    this.context = context;
+    this.ruby = ruby;
+    this.selector = {
+      scheme: "file",
+      language: "ruby",
+    };
+
+    this.serverStatus = this.createStatusItem(
+      "serverStatus",
+      "Ruby LSP: Starting...",
+      vscode.LanguageStatusSeverity?.Warning
+    );
+
+    this.serverStatus.command = {
+      title: "Restart server",
+      command: ServerCommand.Error,
+    };
+
+    this.createYjitStatus(this.ruby);
+    this.createRubyStatus(this.ruby);
+  }
+
+  public async updateStatus(status: ServerCommand) {
+    if (this.serverStatus) {
+      this.serverStatus.dispose();
+    }
+
+    switch (status) {
+      case ServerCommand.Start: {
+        this.serverStatus = this.createStatusItem(
+          "serverStatus",
+          "Ruby LSP: Running...",
+          vscode.LanguageStatusSeverity?.Warning
+        );
+
+        this.serverStatus.command = {
+          title: "Stop Server",
+          command: ServerCommand.Stop,
+        };
+
+        this.activateGemOutdatedButton();
+
+        break;
+      }
+      case ServerCommand.Stop: {
+        this.serverStatus = this.createStatusItem(
+          "serverStatus",
+          "Ruby LSP: Stopped",
+          vscode.LanguageStatusSeverity?.Warning
+        );
+
+        this.serverStatus.command = {
+          title: "Start Server",
+          command: ServerCommand.Start,
+        };
+
+        break;
+      }
+      case ServerCommand.Error: {
+        this.serverStatus = this.createStatusItem(
+          "serverStatus",
+          "Ruby LSP: Error",
+          vscode.LanguageStatusSeverity?.Error
+        );
+
+        this.serverStatus.command = {
+          title: "Restart Server",
+          command: ServerCommand.Error,
+        };
+      }
+    }
+  }
+
+  public async installGems(): Promise<boolean> {
```

Comment:
Question about the UX for everybody: before we had the "gems are not installed" dialogue, people use to complain regularly that Sorbet or the Ruby LSP weren't working because they forgot to bundle install - which we haven't heard anymore since the dialogue is in place.

Do we risk falling back into that scenario by putting this in the language status item? Should we treat bundle install as the only special case in which we display a dialogue and put everything else in status items?

Reply:
I don't think I'm 100% on this opinion, but I'm leaning towards wanting it in the status item. Thinking about the Sorbet LSP, I like that it doesn't spam me, but I know where to look for feedback if Sorbet is not working. Plus the status being displayed in the status item makes it easy to see if something weird is going on.

Reply:
After discussing this in our team weekly sync, we decided it would be best to have the "gems are not installed" dialogue as a language status item for this PR and then in a future PR, especially if people begin having issues, investigate whether it would be better as a status bar item or revert it back to an information message.